### ISSUE TITLE
Python: Mutual TLS GRPC channel fails TLS handshake with "Illegal Parameter"

### ISSUE THREAD

**Issue by xjtian** [grpc/grpc#23235]
### What version of gRPC and what language are you using?

- Python `grpcio==1.29.0`

### What operating system (Linux, Windows,...) and version?

- Debian Stretch 4.9.0-9-amd64
- Vagrant VM (Virtualbox driver) on MacOS host

### What runtime / compiler are you using (e.g. python version or version of gcc)

- Python 3.5.3

### What did you do?

I'm trying to connect to a GRPC server behind an Nginx ingress which terminates TLS and requires client authentication (mTLS client certificate). I use the following block of Python code to establish the channel (imports hidden for brevity):

```
# Cipher picked from output of openssl s_client below
os.environ['GRPC_SSL_CIPHER_SUITES'] = 'ECDHE-RSA-AES256-GCM-SHA384'

rootca = open('/var/opt/magma/tmp/certs/rootCA.pem', 'rb').read()
cert = open('/var/opt/magma/certs/gateway.crt', 'rb').read()
key = open('/var/opt/magma/certs/gateway.key', 'rb').read()
ssl_creds = grpc.ssl_channel_credentials(root_certificates=rootca, private_key=key, certificate_chain=cert)

channel = grpc.secure_channel('controller-staging.magma.etagecom.io:443', ssl_creds, options=[('grpc.default_authority', 'metricsd-controller-staging.magma.etagecom.io')])
client = MetricsControllerStub(channel)
res = client.Push(PushedMetricsContainer(networkId='foobar'))
```

For reference, this is the relevant Nginx config on the server side:

```
  server {
    listen              443 ssl http2;
    server_name         ~^(?<srv>.+)-controller-staging.magma.etagecom.io$;

    ssl_certificate     /var/opt/magma/certs/controller.crt;
    ssl_certificate_key /var/opt/magma/certs/controller.key;
    ssl_verify_client on;
    ssl_client_certificate /var/opt/magma/certs/certifier.pem;

    location / {
      grpc_pass grpc://backend:$srvport;
    }
  }
```

### What did you expect to see?

- TLS handshake success and a successful RPC call. 

### What did you see instead?

TLS handshake fails with Illegal Parameter:

```
I0617 06:26:27.854170917   23047 ssl_transport_security.cc:222]      HANDSHAKE START -       TLS client start_connect  - !!!!!!
I0617 06:26:27.854282536   23047 ssl_transport_security.cc:222]                 LOOP -    TLS client enter_early_data  - !!!!!!
I0617 06:26:27.854296264   23047 ssl_transport_security.cc:222]                 LOOP -   TLS client read_server_hello  - !!!!!!
I0617 06:26:28.040302860   23047 ssl_transport_security.cc:222]                 LOOP - TLS client read_server_certifi  - !!!!!!
I0617 06:26:28.040768920   23047 ssl_transport_security.cc:222]                 LOOP - TLS client read_certificate_st  - !!!!!!
I0617 06:26:28.040795335   23047 ssl_transport_security.cc:222]                 LOOP - TLS client verify_server_certi  - !!!!!!
I0617 06:26:28.041067438   23047 ssl_transport_security.cc:222]                 LOOP - TLS client read_server_key_exc  - !!!!!!
I0617 06:26:28.041166825   23047 ssl_transport_security.cc:222]                 LOOP - TLS client read_certificate_re  - !!!!!!
I0617 06:26:28.041189527   23047 ssl_transport_security.cc:222]                 LOOP - TLS client read_server_hello_d  - !!!!!!
I0617 06:26:28.041195069   23047 ssl_transport_security.cc:222]                 LOOP - TLS client send_client_certifi  - !!!!!!
I0617 06:26:28.041219760   23047 ssl_transport_security.cc:222]                 LOOP - TLS client send_client_key_exc  - !!!!!!
I0617 06:26:28.041821004   23047 ssl_transport_security.cc:222]                 LOOP - TLS client send_client_certifi  - !!!!!!
I0617 06:26:28.043042052   23047 ssl_transport_security.cc:222]                 LOOP - TLS client send_client_finishe  - !!!!!!
I0617 06:26:28.043069409   23047 ssl_transport_security.cc:222]                 LOOP -       TLS client finish_flight  - !!!!!!
I0617 06:26:28.043080088   23047 ssl_transport_security.cc:222]                 LOOP - TLS client read_session_ticket  - !!!!!!
E0617 06:26:28.230393921   23047 ssl_transport_security.cc:1379] Handshake failed with fatal error SSL_ERROR_SSL: error:10000417:SSL routines:OPENSSL_internal:SSLV3_ALERT_ILLEGAL_PARAMETER.
D0617 06:26:28.230430352   23047 security_handshaker.cc:184] Security handshake failed: {"created":"@1592375188.230421755","description":"Handshake failed","file":"src/core/lib/security/transport/security_handshaker.cc","file_line":307,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
```

On the server side, I get this log from nginx:

```
[crit] 8#8: *113777 SSL_do_handshake() failed (SSL: error:1414D17A:SSL routines:tls12_check_peer_sigalg:wrong curve) while SSL handshaking
```

### Anything else we should know about your project / environment?

- The private key is an EC key, not RSA
- Using an http2 reverse proxy configured to open a TLS session with the backend server in between the Python GRPC client and the Nginx ingress still works for establishing connections.
- `openssl s_client` verifies that my client cert and key check out with the server:

```
$ openssl s_client -CAfile /var/opt/magma/tmp/certs/rootCA.pem -key /var/opt/magma/certs/gateway.key -cert /var/opt/magma/certs/gateway.crt -connect controller-staging.magma.etagecom.io:443

CONNECTED(00000003)
...
Requested Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:0x07+0x08:0x08+0x08:0x09+0x08:0x0A+0x08:0x0B+0x08:0x04+0x08:0x05+0x08:0x06+0x08:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:RSA+SHA224:DSA+SHA224:DSA+SHA256:DSA+SHA384:DSA+SHA512
Shared Requested Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:RSA+SHA224:DSA+SHA224:DSA+SHA256:DSA+SHA384:DSA+SHA512
...
---
SSL handshake has read 4281 bytes and written 1056 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
...
    Verify return code: 0 (ok)
    Extended master secret: yes
```

- I took a short packet capture while running my Python script: [tlscap.pcap.txt](https://github.com/grpc/grpc/files/4790797/tlscap.pcap.txt). The first TLS handshake starts at frame 16 and 41 has the Illegal Parameter error. Filtering `!ssh` in Wireshark cleans up the noise.

- Some additional context on the server side:

```
$ nginx -V
nginx version: nginx/1.17.10
built by gcc 8.3.0 (Debian 8.3.0-6)
built with OpenSSL 1.1.1d  10 Sep 2019
TLS SNI support enabled

$ openssl ciphers
TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES256-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:RSA-PSK-AES256-GCM-SHA384:DHE-PSK-AES256-GCM-SHA384:RSA-PSK-CHACHA20-POLY1305:DHE-PSK-CHACHA20-POLY1305:ECDHE-PSK-CHACHA20-POLY1305:AES256-GCM-SHA384:PSK-AES256-GCM-SHA384:PSK-CHACHA20-POLY1305:RSA-PSK-AES128-GCM-SHA256:DHE-PSK-AES128-GCM-SHA256:AES128-GCM-SHA256:PSK-AES128-GCM-SHA256:AES256-SHA256:AES128-SHA256:ECDHE-PSK-AES256-CBC-SHA384:ECDHE-PSK-AES256-CBC-SHA:SRP-RSA-AES-256-CBC-SHA:SRP-AES-256-CBC-SHA:RSA-PSK-AES256-CBC-SHA384:DHE-PSK-AES256-CBC-SHA384:RSA-PSK-AES256-CBC-SHA:DHE-PSK-AES256-CBC-SHA:AES256-SHA:PSK-AES256-CBC-SHA384:PSK-AES256-CBC-SHA:ECDHE-PSK-AES128-CBC-SHA256:ECDHE-PSK-AES128-CBC-SHA:SRP-RSA-AES-128-CBC-SHA:SRP-AES-128-CBC-SHA:RSA-PSK-AES128-CBC-SHA256:DHE-PSK-AES128-CBC-SHA256:RSA-PSK-AES128-CBC-SHA:DHE-PSK-AES128-CBC-SHA:AES128-SHA:PSK-AES128-CBC-SHA256:PSK-AES128-CBC-SHA
```

**Comment by xjtian** [comment]
Additional info: I was able to connect to this GRPC server using a Golang client. The Python client code continues to fail the TLS handshake.

**Comment by jiangtaoli2016** [comment]
gRPC wrapped languages only support P256 curve. Most likely the problem is that you use a different curve.
@xjtian could you print out the `gateway.crt?`

**Comment by xjtian** [comment]
Ah, I see - it's a P384 curve. The golang client library seems to support this just fine, but I do see https://github.com/grpc/grpc/issues/6722 still open r.e. P384 support. Let me retry with a P256 curve and see what happens.

```
$ openssl x509 -in gateway.crt -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            20:58:db:3e:12:54:7d:93:93:e3:b6:94:89:9d:be:2d
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = US, CN = certifier.magma.test
        Validity
            Not Before: Jun 23 00:45:41 2020 GMT
            Not After : Jun 27 01:45:41 2020 GMT
        Subject: CN = c80398d6-5e05-4ff0-a3ec-c0fdd6e0fbc5
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (384 bit)
                pub:
                    04:2c:07:dc:65:02:99:35:a1:ca:51:aa:39:be:d0:
                    9f:53:34:6a:43:72:6e:0c:9c:ab:87:f2:2c:93:5f:
                    ee:e9:b9:c8:55:92:ad:54:c8:1a:06:7d:59:73:0e:
                    04:05:96:f8:d7:7b:06:60:a6:04:08:7a:98:11:e6:
                    01:b1:ef:39:6d:64:a8:80:22:d5:98:08:64:1a:a7:
                    9f:40:72:a0:a1:13:05:69:f4:6c:11:1a:33:de:7e:
                    8f:7f:63:60:77:7d:e7
                ASN1 OID: secp384r1
                NIST CURVE: P-384
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Client Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Authority Key Identifier:
                keyid:5D:E0:6B:EA:1E:26:9B:E5:0A:32:C5:6C:BF:CD:D0:45:21:3A:17:17

    Signature Algorithm: sha256WithRSAEncryption
         19:3b:14:46:3c:e6:a2:ed:37:d8:0e:6d:11:05:05:2e:ed:ab:
         d5:ed:04:97:b7:ff:d1:b5:7c:ca:74:6f:3d:7e:56:49:7c:ce:
         66:d4:32:12:32:55:eb:38:19:39:2d:bc:52:7b:49:c1:ab:e9:
         0f:f5:98:2b:79:b5:58:56:99:29:74:4c:f6:eb:ea:d7:ea:27:
         ef:f5:75:11:f1:f6:07:f6:6f:2e:a4:1c:93:9b:7d:a0:13:99:
         2a:c9:e8:d1:05:5f:10:46:fd:20:94:d7:88:9f:2d:cb:b4:71:
         25:f8:55:6e:6e:75:4f:70:7b:42:a2:67:07:f1:48:76:3e:95:
         0a:c3:70:99:d9:a6:b4:be:2d:d5:0a:34:7b:9f:9c:61:8f:ab:
         de:33:22:98:67:55:73:21:e7:15:f3:1a:8f:b4:c3:88:7a:c0:
         ee:86:3c:97:0d:5e:70:14:ee:1d:14:e9:50:17:e4:f5:34:e5:
         14:12:10:54:77:10:63:6e:82:02:84:02:5c:7b:85:46:0b:da:
         ef:d0:d7:07:64:d6:5a:e7:cc:ca:e2:b7:ae:9e:fd:fb:0d:25:
         02:bb:8e:33:7d:c6:20:c1:4f:1d:45:fc:f6:ed:df:47:76:6e:
         72:b1:39:4e:6a:13:13:f5:d2:5f:68:3f:9f:dc:ac:0d:b9:29:
         b1:68:4d:25
-----BEGIN CERTIFICATE-----
MIICjTCCAXWgAwIBAgIQIFjbPhJUfZOT47aUiZ2+LTANBgkqhkiG9w0BAQsFADAs
MQswCQYDVQQGEwJVUzEdMBsGA1UEAwwUY2VydGlmaWVyLm1hZ21hLnRlc3QwHhcN
MjAwNjIzMDA0NTQxWhcNMjAwNjI3MDE0NTQxWjAvMS0wKwYDVQQDEyRjODAzOThk
Ni01ZTA1LTRmZjAtYTNlYy1jMGZkZDZlMGZiYzUwdjAQBgcqhkjOPQIBBgUrgQQA
IgNiAAQsB9xlApk1ocpRqjm+0J9TNGpDcm4MnKuH8iyTX+7puchVkq1UyBoGfVlz
DgQFlvjXewZgpgQIepgR5gGx7zltZKiAItWYCGQap59AcqChEwVp9GwRGjPefo9/
Y2B3feejVjBUMA4GA1UdDwEB/wQEAwIFoDATBgNVHSUEDDAKBggrBgEFBQcDAjAM
BgNVHRMBAf8EAjAAMB8GA1UdIwQYMBaAFF3ga+oeJpvlCjLFbL/N0EUhOhcXMA0G
CSqGSIb3DQEBCwUAA4IBAQAZOxRGPOai7TfYDm0RBQUu7avV7QSXt//RtXzKdG89
flZJfM5m1DISMlXrOBk5LbxSe0nBq+kP9ZgrebVYVpkpdEz26+rX6ifv9XUR8fYH
9m8upByTm32gE5kqyejRBV8QRv0glNeIny3LtHEl+FVubnVPcHtComcH8Uh2PpUK
w3CZ2aa0vi3VCjR7n5xhj6veMyKYZ1VzIecV8xqPtMOIesDuhjyXDV5wFO4dFOlQ
F+T1NOUUEhBUdxBjboIChAJce4VGC9rv0NcHZNZa58zK4reunv37DSUCu44zfcYg
wU8dRfz27d9Hdm5ysTlOahMT9dJfaD+f3KwNuSmxaE0l
-----END CERTIFICATE-----
```

**Comment by xjtian** [comment]
OK, I verified that the Python client can connect to my server when I use a P256 curve for the client cert. Is there any plan for the Python GRPC library to support any other curves like the Golang library for example?

**Comment by jqueuniet** [comment]
Ran into this issue today using python3-grpcio 1.30.2-3 on Debian Buster. In case anyone wonders, only the server/client certs need to be P256 curves, my P521 root and intermediate certificate authorities worked as is.

**Comment by luoxiaohei** [comment]
Is there any progress on this issue currently? In grpcio version 1.54.2, it still only permits the use of P256 elliptic curve client certificates.

**Comment by rockspore** [comment]
This should be fixed by #34867. We will support prime256v1, secp384r1 and secp521r1 for TLSv1.2
